### PR TITLE
Support ESLint v7

### PR DESCRIPTION
**Pull request by koba04**
ESLint v7 has been released so we'd like to support ESLint v7.
https://eslint.org/blog/2020/05/eslint-v7.0.0-released

**THIS PR IS A BREAKING CHANGE because this drops to support ESLint v6.**

## Breaking Changes

- Drop support ESLint v6
- Enable new rules in `@typescript-eslint/recommended`
  - [`@typescript-eslint/no-extra-non-null-assertion`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-extra-non-null-assertion.md)
  - [`@typescript-eslint/no-extra-semi`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-extra-semi.md)
  - [`@typescript-eslint/no-non-null-asserted-optional-chain`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-non-null-asserted-optional-chain.md)
  - ⚠️ [`@typescript-eslint/prefer-as-const`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-as-const.md)

## Tasks

- [x] update ESLInt in devDeps and peerDeps
- [x] use ESLint class instead of CLIEngine
- Update all plugins versions that are compatible ESLint v7
  - [x] `@typescript-eslint/eslint-plugin` https://github.com/typescript-eslint/typescript-eslint/issues/1550
  - [x] `@typescript-eslint/parser` https://github.com/typescript-eslint/typescript-eslint/issues/1550
  - [x] `eslint-plugin-flowtype` https://github.com/gajus/eslint-plugin-flowtype/issues/446
  - [x] `eslint-plugin-import ` https://github.com/benmosher/eslint-plugin-import/commit/6a110dd16a7fd775f08601054bf14ffd503eea7b
  - [x] `eslint-plugin-jsx-a11y` https://github.com/evcohen/eslint-plugin-jsx-a11y/issues/683
  - [x] `eslint-plugin-node` https://github.com/mysticatea/eslint-plugin-node/issues/223
  - [x] `eslint-plugin-prettier` https://github.com/prettier/eslint-plugin-prettier/pull/291
  - [x] `eslint-plugin-react` https://github.com/yannickcr/eslint-plugin-react/pull/2635
  - [x] `eslint-plugin-react-hooks` https://github.com/facebook/react/pull/18878

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,20 +37,20 @@
     "typescript": "^3.9.3"
   },
   "peerDependencies": {
-    "eslint": "^6.4.0",
-    "typescript": "^3.3.3333"
+    "eslint": "^7.0.0",
+    "typescript": ">=3.3.1 <3.10.0"
```
Comment (toshi-toma) [inline package.json:41]: This versions referenced to [typescript-eslint#supported-typescript-version](https://github.com/typescript-eslint/typescript-eslint#supported-typescript-version)

Comment (koba04) [reply package.json]: @toshi-toma Could you suggest this again because of stale?

File: test/base-test.js [hunk 2]
```diff
@@ -2,15 +2,13 @@ const assert = require("assert");
...
+  it("should get expected errors and warninigs with base config", async () => {
+    const result = await runLintWithFixtures("base");
     assert.deepStrictEqual(result, {
       "error.js": {
         errors: [
           "no-var",
           "no-unused-vars",
-          "no-var",
-          "no-redeclare",
           "getter-return",
```
Comment (koba04) [inline test/base-test.js:12]: The order of the errors is not the same between Node v10 and v12, so I temporarily disable the errors.

File: test/es5-test.js [hunk 3]
```diff
@@ -2,11 +2,11 @@ const assert = require("assert");
 const runLintWithFixtures = require("./lib/runLintWithFixtures");
 
 describe("es5", () => {
-  it("should get expected errors and warninigs with es5 config", () => {
-    const result = runLintWithFixtures("es5");
+  it("should get expected errors and warninigs with es5 config", async () => {
+    const result = await runLintWithFixtures("es5");
     assert.deepStrictEqual(result, {
       "error.js": {
-        errors: ["no-unused-vars", "no-redeclare"]
```
Comment (koba04) [inline test/es5-test.js:9]: The order of the errors is not the same between Node v10 and v12, so I temporarily disable the errors.

**Comment by koba04** [comment]
`eslint-plugin-node` seems to work fine with ESLint v7, so I'll be able to merge this.

File: package.json [hunk 4]
```diff
@@ -37,20 +37,20 @@
     "typescript": "^3.9.5"
   },
   "peerDependencies": {
-    "eslint": "^6.4.0",
+    "eslint": "^7.0.0",
     "typescript": "^3.3.3333"
```
Comment (toshi-toma) [inline package.json:41]:
```suggestion
    "typescript": ">=3.3.1 <3.10.0"
```

File: package.json [hunk 5]
```diff
@@ -37,20 +37,20 @@
...
-    "@typescript-eslint/parser": "^2.34.0",
+    "@typescript-eslint/eslint-plugin": "^3.2.0",
+    "@typescript-eslint/parser": "^3.2.0",
     "babel-eslint": "^10.1.0",
     "eslint-config-prettier": "^6.10.1",
-    "eslint-plugin-flowtype": "^4.7.0",
-    "eslint-plugin-import": "^2.20.2",
+    "eslint-plugin-flowtype": "^5.1.3",
+    "eslint-plugin-import": "^2.21.2",
     "eslint-plugin-jsx-a11y": "6.2.3",
```
[the remark this answers is not in the document]
Comment (koba04) [reply package.json]: @toshi-toma Thanks. I've updated it to the latest version.

**Pull request merged**